### ISSUE TITLE
EvalError: call to eval() blocked by CSP

### ISSUE THREAD

**Issue by elliot-sawyer** [silverstripe/silverstripe-mfa#358]
If you run MFA on a server with a content security policy that disallows `unsafe-eval` statements, the application will refuse to load because the compiled application is invoking an `eval` command. The only way around it is to compile the app without an `eval` or allow `unsafe-eval` in your policy.

Steps to reproduce (fresh install of silverstripe/mfa: 4.0.1):
1. Configure an account with MFA and the security key (aka Yubikey) option
2. Install a CSP with the following header: `X-Content-Security-Policy: base-uri 'none'; default-src 'none'; child-src; connect-src 'none'; font-src 'self'; form-action 'self'; frame-ancestors 'none'; img-src 'self' blob: data:; media-src 'none'; object-src 'none'; script-src 'self'; style-src 'self'; report-uri /csp_violation_reporting_endpoint; report-to /csp_violation_reporting_endpoint; upgrade-insecure-requests`
3. Go to the login screen and attempt to log in with your username/pass. (the logo is also obscured due to CSP violations)
4. Observe: the instructions for logging in with the security key are not displayed. The Yubikey input is ignored. The above error is displayed in the console.

**Comment by Firesphere** [comment]
For security reasons, I would call this an `impact/high`. Literally not being able to use a Secure Policy on your _login_ forms is quite a thing, `unsafe-eval` would render the javascript execution protection useless, meaning _any_ plugin in the browser, or XSS could be executed.

**Comment by elliot-sawyer** [comment]
The origins of the eval statements are in `vendor/silverstripe/mfa/client/dist/js/injector.js`. This seems to be a pre-compiled library?

`vendor/silverstripe/mfa/client/dist/js/bundle.js`

bundle.js also contains a few snippets that will violate the CSP. You'll need to look out for Javascript functions that evaluate strings as code, such as these ones (Ctrl F for `Function("`)

```
Function("C,a","
Function("return this")()}).
Function("return this")()}
Function("return this")().
Function("r","regeneratorRuntime
Function("return this")()||(
```

**Comment by elliot-sawyer** [comment]
The eval statements are removed from the injector library when you remove this line from webpack.config.js: 

`devtool: 'eval-source-map',`

Compile with `yarn build` and copy the generated dist file to the mfa project. The CSP no longer complains about the eval statement. 

For anyone else looking to get this to working with a strict CSP, a few other things need to be allowed. Note this is using JSON syntax for readability
```
//to allow the Yubikey
  "connect-src": {
    "self": true
  },

// to load the React app. Avoidable if you nonce or add an integrity attribute
  "script-src": {
    "unsafe-inline": true,
  },

  //to display the SVG logo and MFA instructions. Avoidable if you nonce or add an integrity attribute
  "style-src": {
    "unsafe-inline": true,
  },
```

**Comment by elliot-sawyer** [comment]
A project maintainer will need to fix this. While I could supply a new `dist/injector.js` file, nobody should trust a dist file supplied by a non-maintainer. A better approach would be to remove this hardcoded dist entirely, and build from source using webpack

**Comment by elliot-sawyer** [comment]
Pull request here: https://github.com/silverstripe/react-injector/pull/6

Since injector.js is hardcoded into the project, you'll need to re-compile and copy to your project
```
(react-injector) yarn
(react-injector) yarn build
(react-injector) cp dist/injector.js /path/to/project/vendor/silverstripe/mfa/client/dist/js/injector.js
(silverstripe/mfa) composer vendor-expose
```

**Comment by NightJar** [comment]
Thanks for following up on this @elliot-sawyer

**Comment by ScopeyNZ** [comment]
> A project maintainer will need to fix this. While I could supply a new `dist/injector.js` file, nobody should trust a dist file supplied by a non-maintainer. A better approach would be to remove this hardcoded dist entirely, and build from source using webpack

Just FYI - the webpack build process of this library compiles the injector from source - it's not copied. After the fix is merged into the upstream repo and published on NPM there will just need to be a follow up here that `yarn install`'s and `yarn build`'s.

**Comment by elliot-sawyer** [comment]
You're right! I just didn't see a webpack config in this project so assumed it was hardcoded in there. I see now that package.json is checking out a specific tag of react-injector. Once that change is merged the tag can just get updated.

**Comment by ScopeyNZ** [comment]
Our CI should catch nefarious dist files too - as the CI scripts will `yarn install` and `yarn build` just to compare with what's committed.

**Comment by Cheddam** [comment]
@elliot-sawyer This should be resolved now! Feel free to reopen if you're still experiencing issues.

**Comment by elliot-sawyer** [comment]
@Cheddam I think it's still an issue because [the dist is still shipped with 4.0.1](https://github.com/silverstripe/silverstripe-mfa/blob/4.0.1/client/dist/js/injector.js). Per @ScopeyNZ 's suggestion above, I think someone needs to `yarn install` and `yarn build`, then tag with 4.0.2?

**Comment by Cheddam** [comment]
@elliot-sawyer The rebuild happened in #361, but there isn't a tagged release yet - I'll sort that shortly.

**Comment by phptek** [comment]
@Cheddam Has this patch made it into a release yet? I don't see a merge commit for it under "commits" for 4.0.2. Thanks!

**Comment by Cheddam** [comment]
@phptek Yep, 4.0.2 covers it. The change is listed in the release, but the specified commit hash is for the change commit itself, rather than the related merge commit.

**Comment by Nilsheated** [comment]
Here’s a quick answer https://stackdev.space/questions/invoke-the-child-frames-return-operation

**Comment by andreashochhaltervisitrans** [comment]
Still an issue in 6.0.4? we have 2026?

**Comment by NightJar** [comment]
No. This issue (#358) was fixed in 2019.
If you have found a similar problem or a regression then you should open a new issue _politely_ explaining the details.
